pkg/controller/common/cleaner_test.go

```diff
@@ -237,3 +239,110 @@ func AssertAction(t *testing.T, got []k8stesting.Action, expected k8stesting.Act
 
 	require.True(t, found, "expected to find action", expected)
 }
+
+func TestNewCleaner(t *testing.T) {
+	m := getManager()
+
+	err := NewCleaner(m, "test-new-cleaner", RetrieverEmpty())
+	require.NoError(t, err)
+}
+
+func TestClean(t *testing.T) {
+	d := dynamicfake.NewSimpleDynamicClientWithCustomListKinds(runtime.NewScheme(), map[schema.GroupVersionResource]string{
+		namespacedGvr:    namespacedList,
+		nonNamespacedGvr: nonNamespacedList,
+	})
+	m := getManager()
+	mapper, err := apiutil.NewDynamicRESTMapper(m.GetConfig(), apiutil.WithLazyDiscovery)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name           string
+		c              *cleaner
+		expectedErrMsg string
+	}{
+		{
+			name: "nil retriver",
+			c: &cleaner{
+				name:      "test-name",
+				dynamic:   d,
+				clientset: fakeClientset(),
+				logger:    logr.Discard(),
+			},
+			expectedErrMsg: "retriever is nil",
+		},
+		{
+			name: "rest mapper err",
+			c: &cleaner{
+				name:      "test-name",
+				dynamic:   d,
+				clientset: fakeClientset(),
+				logger:    logr.Discard(),
+				mapper:    mapper,
+				retriever: RetrieverEmpty(),
+			},
+			expectedErrMsg: "",
+		},
+	}
+
+	for _, test := range tests {
+		err := test.c.Clean(context.Background())
+		if err != nil {
+			require.Equal(t, test.expectedErrMsg, err.Error(), "should return expected error msg")
+		}
+	}
+}
+
+func TestCleanerStart(t *testing.T) {
+	d := dynamicfake.NewSimpleDynamicClientWithCustomListKinds(runtime.NewScheme(), map[schema.GroupVersionResource]string{
+		namespacedGvr:    namespacedList,
+		nonNamespacedGvr: nonNamespacedList,
+	})
+	m := getManager()
+	mapper, err := apiutil.NewDynamicRESTMapper(m.GetConfig(), apiutil.WithLazyDiscovery)
+	require.NoError(t, err)
+
+	tests := []struct {
+		name           string
+		c              *cleaner
+		expectedErrMsg string
+	}{
+		{
+			name: "failing to start",
+			c: &cleaner{
+				name:       "test-name",
+				dynamic:    d,
+				clientset:  fakeClientset(),
+				logger:     logr.Discard(),
+				maxRetries: 1,
+			},
+			expectedErrMsg: "retriever is nil",
+		},
+		{
+			name: "starts cleaner",
+			c: &cleaner{
+				name:       "test-name",
+				dynamic:    d,
+				clientset:  fakeClientset(),
+				logger:     logr.Discard(),
+				mapper:     mapper,
+				retriever:  RetrieverEmpty(),
+				maxRetries: 1,
+			},
+			expectedErrMsg: "",
+		},
+	}
+	for _, test := range tests {
+		err := test.c.Start(context.Background())
+		if err != nil {
+			require.Equal(t, test.expectedErrMsg, err.Error(), "should return expected error msg")
+		}
+	}
+}
+
+func getManager() manager.Manager {
+	testenv := &envtest.Environment{}
+	cfg, _ := testenv.Start()
```

Comment:
What is the result of never calling `testenv.End()`? I'm curious if that leaves a k8s environment hanging or if they are automatically cleaned at the end of the Go process?

I also wonder if we need a separate environment for each test or if it should be a global one. Most of the examples i see are global ones.

Reply:
I checked what `testenv.Stop()` does:
`Stop stops a running server. Previously installed CRDs, as listed in CRDInstallOptions.CRDs, will be uninstalled if CRDInstallOptions.CleanUpAfterUse are set to true` So not calling that function leaves env hanging. We should stick to global env.